// @blisp/reader/number/read-number.js
const peekChar = require("../peek-char")
const peekCharCode = require("../peek-char-code")
const readChar = require("../read-char")
const readTable = require("./number-read-table")

const ZERO = "0".charCodeAt(0)
const ONE = "1".charCodeAt(0)
const SEVEN = "7".charCodeAt(0)
const NINE = "9".charCodeAt(0)
const A = "A".charCodeAt(0)
const F = "F".charCodeAt(0)
const a = "a".charCodeAt(0)
const f = "f".charCodeAt(0)

function inRange(stream, start, stop) {
  const charCode = peekCharCode(stream)
  return start <= charCode && charCode <= stop
}

function isBinaryDigit(stream) {
  return inRange(stream, ZERO, ONE)
}

function isOctalDigit(stream) {
  return inRange(stream, ZERO, SEVEN)
}

function isDecimalDigit(stream) {
  return inRange(stream, ZERO, NINE)
}

function isHexDigit(stream) {
  return (
    isDecimalDigit(stream) || inRange(stream, A, F) || inRange(stream, a, f)
  )
}

function readDigits(test) {
  return (stream, char, number) => {
    reader = readTable[char]
    while (reader || test(stream)) {
      number = reader ? reader(stream, char, number) : number + readChar(stream)
      reader = readTable[peekChar(stream)]
    }
    return number
  }
}

module.exports = function readNumber(stream, char) {
  let number = ""
  let test = isDecimalDigit
  if (char === "0") {
    number += readChar(stream)
    char = peekChar(stream)
    switch (peekChar(stream)) {
      case "b":
      case "B":
        test = isBinaryDigit
        number += readChar(stream)
        char = peekChar(stream)
        break

      case "o":
      case "O":
        test = isOctalDigit
        number += readChar(stream)
        char = peekChar(stream)
        break

      case "x":
      case "X":
        test = isHexDigit
        number += readChar(stream)
        char = peekChar(stream)
        break
    }
    if (test === isDecimalDigit && isDecimalDigit(stream)) {
      throw new Error(`Invalid input expected '.' got '${char}'`)
    }
  }

  number = readDigits(test)(stream, char, number)

  if (test === isDecimalDigit) {
    char = peekChar(stream)
    if (char === ".") {
      number += readChar(stream)
      char = peekChar(stream)
      number = readDigits(test)(stream, char, number)
      char = peekChar(stream)
    }
    if (char === "e" || char === "E") {
      number += readChar(stream)
      char = peekChar(stream)
      if (char === "-" || char === "+") {
        number += readChar(stream)
        char = peekChar(stream)
      }
      number = readDigits(test)(stream, char, number)
    }
  }
  char = peekChar(stream)
  if (!(stream.eof() || isWhitespace(char))) {
    throw new Error(`Unexpected input ${char}`)
  }
  return Number(number)
}
